app/models/theme.rb

```diff
@@ -434,25 +429,71 @@ def self.refresh_message_for_targets(targets, theme_ids)
   end
 
   def self.resolve_baked_field(theme_ids, target, name)
-    if target == :extra_js
-      require_rebake =
-        ThemeField.where(theme_id: theme_ids, target_id: Theme.targets[:extra_js]).where(
-          "compiler_version <> ?",
-          Theme.compiler_version,
-        )
-      require_rebake.each { |tf| tf.ensure_baked! }
-      require_rebake
-        .map(&:theme_id)
-        .uniq
-        .each { |theme_id| Theme.find(theme_id).update_javascript_cache! }
-      caches = JavascriptCache.where(theme_id: theme_ids)
-      caches = caches.sort_by { |cache| theme_ids.index(cache.theme_id) }
-      return caches.map { |c| <<~HTML.html_safe }.join("\n")
+    target = target.to_sym
+    name = name&.to_sym
+
+    target = :mobile if target == :mobile_theme
+    target = :desktop if target == :desktop_theme
+
+    case target
+    when :extra_js
+      get_set_cache("#{theme_ids.join(",")}:extra_js:#{Theme.compiler_version}") do
+        require_rebake =
+          ThemeField.where(theme_id: theme_ids, target_id: Theme.targets[:extra_js]).where(
+            "compiler_version <> ?",
+            Theme.compiler_version,
+          )
+
+        ActiveRecord::Base.transaction do
+          require_rebake.each { |tf| tf.ensure_baked! }
+
+          Theme.where(id: require_rebake.map(&:theme_id)).each(&:update_javascript_cache!)
+        end
+
+        caches =
+          JavascriptCache
+            .where(theme_id: theme_ids)
+            .index_by(&:theme_id)
+            .values_at(*theme_ids)
+            .compact
+
+        caches.map { |c| <<~HTML.html_safe }.join("\n")
           <link rel="preload" href="#{c.url}" as="script">
           <script defer src='#{c.url}' data-theme-id='#{c.theme_id}'></script>
         HTML
+      end
+    when :translations
+      theme_field_values(theme_ids, :translations, I18n.fallbacks[name])
+        .values
+        .compact
+        .first
+        &.first
+        &.join("\n") || ""
+    else
+      theme_field_values(theme_ids, [:common, target], name).values.compact.flatten.join("\n")
+    end
+  end
+
+  def self.theme_field_values(theme_ids, targets, names)
+    cache.defer_get_set_bulk(
+      Array(theme_ids).product(Array(targets), Array(names)),
+      lambda do |(theme_id, target, name)|
+        "#{theme_id}:#{target}:#{name}:#{Theme.compiler_version}"
+      end,
+    ) do |keys|
+      keys.map! { |theme_id, target, name| [theme_id, Theme.targets[target], name.to_s] }
```

Comment:
I'm wondering if this is intentional where the keys being generated based on the inputs to `defer_get_set_bulk` can be mutated by the block.

Reply:
Oh, good catch.